### PR TITLE
populate "sources" in packit.yaml

### LINKED ISSUE #179: Save the URL of the source-archives in `.packit.yaml`
Instead of commiting the source-archives when converting/updating a source-git repo from a dist-git repo, create a `sources` entry in `.packit.yaml` for each archive mentioned in the metadata file (`.<package>.metadata`), specifying the download URL of those archives.

Note that `packit srpm` currently expects source-archives to be found in the `SPECS` sub-directory.

For example, when processing the `c8s` branch of `rsync`:

```
$ cat .rsync.metadata
82e7829c0b3cefbd33c233005341e2073c425629 SOURCES/rsync-3.1.3.tar.gz
74c16510a18ef43d797f9ceba6150f0862568cc0 SOURCES/rsync-patches-3.1.3.tar.gz
```

Based on this `.packit.yaml` should contain:

```yaml
sources:
  - path: SPECS/rsync-3.1.3.tar.gz
    url: https://git.centos.org/sources/rsync/c8s/82e7829c0b3cefbd33c233005341e2073c425629
  - path: SPECS/rsync-3.1.3.tar.gz
    url: https://git.centos.org/sources/rsync/c8s/74c16510a18ef43d797f9ceba6150f0862568cc0
```

This should be done after packit/packit#1105 is implemented.

### PR DESCRIPTION
**Pull request by TomasTomecek**
Fixes #179

* [x] wait for packit 0.27:
    ```
    We need this for the current PR but it fails to be installed in Zuul - it times out on this task
    let's make the tests fail then and wait for an actual packit release
    ```

### CODE REVIEW

**Comment by jpopelka** [comment]
recheck

**Comment by TomasTomecek** [comment]
Dang, PyPI having issues today:
```
2021-03-04 09:11:36.546423 | test-node | WARNING: Running pip install with root privileges is generally not a good idea. Try `pip3 install --user` instead.
2021-03-04 09:11:36.546444 | test-node |   Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='files.pythonhosted.org', port=443): Read timed out. (read timeout=15)",)': /packages/b1/5c/018bf9a5c24343a664deaea70e61f33f53bb1bd3caf193110f827bfd07e2/sentry_sdk-0.19.5-py2.py3-none-any.whl
2021-03-04 09:11:36.546474 | test-node |   Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/certifi/
```

File: dist2src/core.py [hunk 1]
```diff
@@ -596,6 +628,8 @@ def add_packit_config(self, upstream_ref: str, commit: bool = False):
                 },
             ],
         }
+        self.set_sources_in_packit_config(config)
```
Comment (csomh) [inline dist2src/core.py:631]: Something like: 

```suggestion
            ],
            "sources": self.get_sources()
        }
```

would be probably easier to read, and tell what the content of `.packit.yaml` is going to be.

File: dist2src/core.py [hunk 2]
```diff
@@ -573,6 +590,21 @@ def convert_single_commit(self, origin_branch: str, dest_branch: str):
...
+        for sha, path in self.lookaside_sources():
+
+            # get_sources.sh does git magic in the repo to determine the correct branch
+            for branch in ("c8s", "c8"):
+                url = (
+                    f"https://git.centos.org/sources/{self.package_name}/{branch}/{sha}"
+                )
+                response = requests.head(url)
+                if not response.ok:
+                    continue
```
Comment (csomh) [inline dist2src/core.py:604]: In both places where `add_packit_config` is called, there is an `origin_branch` variable available to store which branch is being converted. Couldn't this be passed down to this function and used, instead of this probing mechanism?

File: dist2src/core.py [hunk 3]
```diff
@@ -611,11 +645,19 @@ def copy_all_sources(self, with_patches: bool = False):
...
             sources += (x.path for x in self.dist_git_spec.get_patches())
 
         for source in sources:
+            # sources are absolute paths as str, lookaside_paths are relative within the repo
+            relative = Path(source).relative_to(self.dist_git_path)
+            if str(relative) in lookaside_source_paths:
+                logger.debug(
+                    f"Source {source} will be fetched from lookaside cache, skipping."
+                )
+                continue
```
Comment (csomh) [inline dist2src/core.py:660]: Why is this required? Once `sources` is set in `.packit.yaml`, the download links for all source archives needed for a build should be available. Which source-archives wouldn't be included in this list that would be fetched by `get_sources.sh` otherwise?

My expectation was, that once we are able to save URLs to the lookaside cache artefacts, we could remove `copy_all_sources()` entirely.

File: dist2src/core.py [hunk 4]
```diff
@@ -271,6 +273,21 @@ def dist_git_spec(self):
         )
         return self._dist_git_spec
 
+    @lru_cache()
```
[the remark this answers is not in the document]
Comment (TomasTomecek) [reply dist2src/core.py:276]: so that we would not load the file twice since it's not gonna change within one session

File: dist2src/core.py (hunk 2 above)
Comment (TomasTomecek) [reply dist2src/core.py:604]: sadly, not; there are cases where the sources are stored on a different branch; example: we are processing c8s but the sources are on the c8 branch - there is some nasty git-magic in get_sources.sh to figure out the correct url

File: dist2src/core.py (hunk 1 above)
Comment (TomasTomecek) [reply dist2src/core.py:631]: oh, yeah, that's nice

File: dist2src/core.py (hunk 3 above)
Comment (TomasTomecek) [reply dist2src/core.py:660]: There are sources which are directly committed in dist-git - usually config files for services or for the whole distro, e.g. in systemd:
* https://git.centos.org/rpms/systemd/blob/c8s/f/SOURCES (scroll to the end)
* https://git.centos.org/rpms/systemd/blob/c8s/f/SPECS/systemd.spec#_36

File: dist2src/core.py (hunk 2 above)
Comment (csomh) [reply dist2src/core.py:604]: Do you have an example for such a case? I wasn't able to find one. Had a look at [rsync](https://git.centos.org/sources/rsync/), but there I see that the same sources (same hash) show up in all the branches between `c8-beta` and `c8s`, which tells me that the lookaside cache does follow whatever branching is happening in Git.

The "magic" `get_sources.sh` does, is not "magic" at all: it just looks up all the branches the current HEAD belongs to, and tries to download the sources using all these branch names, until one of them succeeds. Which kinda is  similar to this implementation.

But, this mechanism [was introduced](https://git.centos.org/centos-git-common/c/c61c1ca3917f2377553df26a5290e5dfb8780631) to `get_sources.sh` in order to handle "detached branches" (meant to be "detached HEAD", I assume). When on a detached HEAD, it does makes sense to follow this approach. But dist2src is expected to always work on the tip of a branch, in which case I don't see why this guessing would be required.

File: dist2src/core.py (hunk 3 above)
Comment (csomh) [reply dist2src/core.py:660]: Meh, what a mess! Totally forgot about this. Thanks for the explanation!

Does there need to be a check to make sure that only files tracked in dist-git are copied and then commited to source-git?

File: dist2src/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (csomh) [reply dist2src/core.py:276]: Okay, I see now, this will cause this method to be run once during the lifetime of the `Dist2Src` object and return the cached return value for subsequent calls. Kinda not obvious, maybe it would worth mentioning in the doc-string.

File: dist2src/core.py (hunk 3 above)
Comment (TomasTomecek) [reply dist2src/core.py:660]: I'd say that's what is happening above.

I'll try to get a list of files in SOURCES/ from git directly and intersect it with the set above and print a warning if something is off and see what our tests would tell.

File: dist2src/core.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (TomasTomecek) [reply dist2src/core.py:276]: Since `@lru_cache` is a standard python decorator with terrible documentation, I'll add there a comment with what you just wrote since it's short and clear, thanks!

File: dist2src/core.py (hunk 2 above)
Comment (TomasTomecek) [reply dist2src/core.py:604]: hmmm, apparently I understood the get_sources.sh code wrong (or things changed with the lookaside in the past couple of months - I clearly remember how someone from the CentOS team was explaining this behaviour to me last year)

I'll drop the magic and see what our test suite will say

thanks for the explanation

**Comment by TomasTomecek** [comment]
seems like PyPI hates us
```
2021-03-10 13:49:02.921490 | test-node | TASK [Install packit with PR#1143] *********************************************
2021-03-10 13:49:02.921521 | test-node | task path: /src/install-deps.yml:42
2021-03-10 14:40:19.887308 | PRE-RUN END RESULT_TIMED_OUT: [untrusted : github.com/packit/dist-git-to-source-git/files/zuul-build-image.yml@main]
2021-03-10 14:40:19.887700 | POST-RUN START: [trusted : github.com/packit/packit-service-zuul/playbooks/base/post.yaml@main]
2021-03-10 14:40:23.821422 | 
2021-03-10 14:40:23.822375 | PLAY [all]
2021-03-10 14:40:23.939206 | 
```

I'm going to drop the part about installing custom packit to see if those tests get going again

File: dist2src/core.py [hunk 5]
```diff
@@ -573,9 +614,34 @@ def convert_single_commit(self, origin_branch: str, dest_branch: str):
...
+        """
+        sources: List[Dict[str, str]] = []
+        if not self.lookaside_sources():
+            # the file can actually be empty
+            return sources
+        for sha, path in self.lookaside_sources():
+            url = f"https://git.centos.org/sources/{self.package_name}/{branch}/{sha}"
+            # packit is too smart, it already expects the archive to be in SPECS
+            # so we just strip the leading SOURCES dir
+            path = path.split("/")[1]
```
Comment (csomh) [inline dist2src/core.py:633]: I'm wondering if it would make sense to do this in `lookaside_sources()` :thinking:

File: tests/conftest.py [hunk 6]
```diff
@@ -14,7 +14,8 @@
     # %autosetup -S git_am + needs https://koji.mbox.centos.org/koji/taginfo?tagID=342
     ("pacemaker", "c8s"),
     ("systemd", "c8s"),  # -S git_am
-    ("kernel", "c8s"),  # !!!
+    # disabling kernel by default for now since it's too big
+    # ("kernel", "c8s"),  # !!!
```
[the remark this answers is not in the document]
Comment (TomasTomecek) [reply tests/conftest.py:18]: I thought you would actually be in favour of this since you complained it takes so long - I don't mind re-enabling it back

[the remark this answers is not in the document]
Comment (TomasTomecek) [reply tests/conftest.py:18]: re-added

[the remark this answers is not in the document]
Comment (csomh) [reply tests/conftest.py:18]: I didn't ask to re-enable it. That was a "finally" face.

File: dist2src/core.py [hunk 7]
```diff
@@ -611,11 +679,24 @@ def copy_all_sources(self, with_patches: bool = False):
         sg_path = self.source_git_path / "SPECS"
         logger.info(f"Copy all sources from {dg_path} to {sg_path}.")
 
-        sources = self.dist_git_spec.get_sources()[:]
+        sources: List[str] = self.dist_git_spec.get_sources()[:]
+        lookaside_source_paths = [path for sha, path in self.lookaside_sources()]
```
[the remark this answers is not in the document]
Comment (TomasTomecek) [reply dist2src/core.py:683]: oh yeah, it looks cleaner now, actually both

File: dist2src/core.py (hunk 5 above)
Comment (TomasTomecek) [reply dist2src/core.py:633]: taking it back, we can't move that because copy_sources wouldn't work since it's using the paths and they need to work

**Comment by TomasTomecek** [comment]
I can safely say that PyPI is the glitch:
```
2021-03-10 16:43:03.429595 | test-node |
2021-03-10 16:43:03.429605 | test-node | Complete!
2021-03-10 16:43:03.429615 | test-node |
2021-03-10 16:43:03.429624 | test-node | TASK [Pip install tools for debugging, tests] **********************************
2021-03-10 16:43:03.429634 | test-node | task path: /src/install-deps.yml:51
2021-03-10 16:46:07.351837 | test-node | fatal: [localhost]: FAILED! => {
2021-03-10 16:46:07.352272 | test-node |     "changed": false,
2021-03-10 16:46:07.352283 | test-node |     "cmd": [
2021-03-10 16:46:07.352288 | test-node |         "/usr/bin/pip3",
2021-03-10 16:46:07.352311 | test-node |         "install",
2021-03-10 16:46:07.352317 | test-node |         "pytest",
2021-03-10 16:46:07.352322 | test-node |         "flexmock"
2021-03-10 16:46:07.352326 | test-node |     ]
2021-03-10 16:46:07.352331 | test-node | }
2021-03-10 16:46:07.352336 | test-node |
2021-03-10 16:46:07.352340 | test-node | MSG:
2021-03-10 16:46:07.352345 | test-node |
2021-03-10 16:46:07.352349 | test-node | stdout: Collecting pytest
2021-03-10 16:46:07.352354 | test-node |   Downloading https://files.pythonhosted.org/packages/a1/cf/7f67585bd2fc0359ec482cf3c430bce3ef6d3f40bc468137225a733e3069/pytest-6.2.2-py3-none-any.whl (280kB)
2021-03-10 16:46:07.352385 | test-node | Collecting flexmock
2021-03-10 16:46:07.352393 | test-node |   Downloading https://files.pythonhosted.org/packages/0d/76/03fc6622d823e6d292c29e444c808788e580b5ad66000f4d75e49c14f51f/flexmock-0.10.4.tar.gz (45kB)
2021-03-10 16:46:07.352398 | test-node | Collecting pluggy<1.0.0a1,>=0.12 (from pytest)
2021-03-10 16:46:07.352415 | test-node |
2021-03-10 16:46:07.352421 | test-node | :stderr: WARNING: Running pip install with root privileges is generally not a good idea. Try `pip3 install --user` instead.
2021-03-10 16:46:07.352457 | test-node |   Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pytest/
2021-03-10 16:46:07.352471 | test-node |   Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pytest/
2021-03-10 16:46:07.352477 | test-node |   Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pytest/
2021-03-10 16:46:07.352482 | test-node |   Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pytest/
2021-03-10 16:46:07.352486 | test-node |   Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pytest/
2021-03-10 16:46:07.352491 | test-node |   Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pluggy/
2021-03-10 16:46:07.352495 | test-node |   Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pluggy/
2021-03-10 16:46:07.352514 | test-node |   Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pluggy/
2021-03-10 16:46:07.352572 | test-node |   Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pluggy/
2021-03-10 16:46:07.352578 | test-node |   Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ReadTimeoutError("HTTPSConnectionPool(host='pypi.python.org', port=443): Read timed out. (read timeout=15)",)': /simple/pluggy/
2021-03-10 16:46:07.352590 | test-node |   Could not find a version that satisfies the requirement pluggy<1.0.0a1,>=0.12 (from pytest) (from versions: )
2021-03-10 16:46:07.352606 | test-node | No matching distribution found for pluggy<1.0.0a1,>=0.12 (from pytest)
2021-03-10 16:46:07.352611 | test-node |
2021-03-10 16:46:07.355200 | test-node |
2021-03-10 16:46:07.355235 | test-node | PLAY RECAP *********************************************************************
2021-03-10 16:46:07.355596 | test-node | localhost                  : ok=4    changed=3    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

**Comment by jpopelka** [comment]
> I can safely say that PyPI is the glitch:

but why that's happening only in this repo (https://github.com/packit/dist-git-to-source-git/pull/186#issuecomment-792890723) is a mystery to me

**Comment by TomasTomecek** [comment]
the CI failure which zuul is about to post is a network flake

**Review by csomh — APPROVED** [review]

**Comment by TomasTomecek** [comment]
I am going to merge despite one test failing and the other one not even starting since I have more changes ready locally and will try to fix current CI problems in the next PR.

@csomh thank you for the reviews!
